### PR TITLE
Laboratory work #1, Anna Beliavskaya - 23FPL2

### PR DESCRIPTION
**Pull request by AnnaBeliavskaya**

### CODE REVIEW

File: lab_1_classify_by_unigrams/start.py [hunk 1]
```diff
@@ -14,9 +14,14 @@ def main() -> None:
         de_text = file_to_read_de.read()
     with open("assets/texts/unknown.txt", "r", encoding="utf-8") as file_to_read_unk:
         unknown_text = file_to_read_unk.read()
-    result = None
+
+    from main import tokenize
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:18]: place import at the top of this file

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -5,8 +5,16 @@
...
 def tokenize(text: str) -> list[str] | None:
+    text = text.lower()
+    cleaned_text = ""
+    for symbol in text:
+        if symbol.isalpha():
+            cleaned_text += symbol
+    tokens = []
+    for token in cleaned_text:
+        tokens += token
+    return tokens
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:17]: place the code after docstrings

File: lab_1_classify_by_unigrams/start.py [hunk 3]
```diff
@@ -14,9 +14,14 @@ def main() -> None:
...
+    result = tokenize(en_text)
+    print(result)
     assert result, "Detection result is None"
 
 
 if __name__ == "__main__":
     main()
+
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:26]: delete whitespaces

**Review by EIzharikov — CHANGES_REQUESTED** [review]
Nice start! I recommend you to hurry up! Good luck with doing this lab work!

File: lab_1_classify_by_unigrams/start.py (hunk 1 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/start.py:18]: done

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/start.py (hunk 1 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/start.py:18]: done

File: lab_1_classify_by_unigrams/start.py (hunk 3 above)
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/start.py:26]: you added whitespace in 26th line

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -34,6 +43,23 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    abs_freq = {}
+
+    if not isinstance(tokens, list):
+        return None
+    total_elements = len(tokens)
+    for token in tokens:
+        if not isinstance(token, str):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:52]: check this at start of func

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -34,6 +43,23 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+            return None
+        else:
+            if token in abs_freq:
+                abs_freq[token] += 1
+            else:
+                abs_freq[token] = 1
+    if total_elements == 0:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:59]: check at the beginning

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -49,7 +75,11 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
-
+    if isinstance(text, str) and isinstance(language, str):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:78]: swap condition

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -49,7 +75,11 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
...
+        preprocessed_text = tokenize(text)
+        relative_frequency = calculate_frequencies(preprocessed_text)
+        language_profile = {"name": language, "freq": relative_frequency}
+        return language_profile
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:82]: return as is

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -49,7 +75,11 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
-
+    if isinstance(text, str) and isinstance(language, str):
+        preprocessed_text = tokenize(text)
+        relative_frequency = calculate_frequencies(preprocessed_text)
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:80]: what if this function will return None?

File: lab_1_classify_by_unigrams/settings.json [hunk 9]
```diff
@@ -1,3 +1,3 @@
 {
-  "target_score": 0
+  "target_score": 6
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/settings.json:2]: I believe that you can achieve 8 at least!

File: lab_1_classify_by_unigrams/start.py [hunk 10]
```diff
@@ -5,6 +5,7 @@
 
 
 def main() -> None:
+    from main import create_language_profile
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:8]: all imports are located at the top of any python file

**Review by EIzharikov — COMMENTED** [review]
Nice progress! Fix comments and achieve 8 at least! You can do it!

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -20,6 +19,16 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if isinstance(text, str):
```
[the remark this answers is not in the document]
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:22]: done

File: lab_1_classify_by_unigrams/start.py (hunk 3 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/start.py:26]: done

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -20,6 +19,16 @@ def tokenize(text: str) -> list[str] | None:
...
+        text = text.lower()
+        cleaned_text = ""
+        for symbol in text:
+            if symbol.isalpha():
+                cleaned_text += symbol
+        tokens = []
+        for token in cleaned_text:
+            tokens += token
```
[the remark this answers is not in the document]
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:30]: += replace with append?

File: lab_1_classify_by_unigrams/start.py (hunk 10 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/start.py:8]: done

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -34,6 +43,23 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    abs_freq = {}
```
[the remark this answers is not in the document]
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:46]: done

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:59]: done

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:78]: done

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:82]: done

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -20,6 +19,18 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(text, str):
+        return None
+    text = text.lower()
+    text = text.replace("º", "")
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:25]: why?

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -35,6 +46,20 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+    abs_freq = {}
+    total_elements = len(tokens)
+    for token in tokens:
+        if not isinstance(token, str):
+            return None
+        if token in abs_freq:
+            abs_freq[token] += 1
+        else:
+            abs_freq[token] = 1
+    rel_freq = {element: freq / total_elements for element, freq in abs_freq.items()}
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:60]: what is rel_freq?  Why are you chosen this name?

File: lab_1_classify_by_unigrams/start.py [hunk 16]
```diff
@@ -14,7 +16,12 @@ def main() -> None:
         de_text = file_to_read_de.read()
     with open("assets/texts/unknown.txt", "r", encoding="utf-8") as file_to_read_unk:
         unknown_text = file_to_read_unk.read()
-    result = None
+
+    english = create_language_profile("en", en_text)
+    german = create_language_profile("en", de_text)
+    unknown = create_language_profile("en", unknown_text)
+    result = detect_language(unknown,english,german)
+    print(result)
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:24]: is it working?

**Review by EIzharikov — COMMENTED** [review]
Good progress! Make CI green, fix comments and be ready for the next review. Good luck!

File: lab_1_classify_by_unigrams/main.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:61]: потому что в словаре относительные частоты символов

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/start.py (hunk 16 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/start.py:28]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:26]: ?

Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py:26]: why are you replacing this?

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py:60]: I do not understand you

File: lab_1_classify_by_unigrams/main.py [hunk 17]
```diff
@@ -20,6 +20,18 @@ def tokenize(text: str) -> list[str] | None:
...
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(text, str):
+        return None
+    text = text.lower()
+    text = text.replace("º", "")
+    cleaned_text = ""
+    for symbol in text:
+        if symbol.isalpha():
+            cleaned_text += symbol
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:30]: can you combine these for loops? you can append symbol in list permanently.

File: lab_1_classify_by_unigrams/main.py [hunk 18]
```diff
@@ -49,6 +74,15 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(text, str) or not isinstance(language, str):
+        return None
+    preprocessed_text = tokenize(text)
+    relative_frequency = calculate_frequencies(preprocessed_text)
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:80]: so here you wrote 'relative_frequency'

File: lab_1_classify_by_unigrams/settings.json [hunk 19]
```diff
@@ -1,3 +1,3 @@
 {
-  "target_score": 0
+  "target_score": 8
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/settings.json:2]: nice aim!

**Review by EIzharikov — COMMENTED** [review]
Nice work!! Fix&answer my small comemnts. make CI green and be ready for lab work defence! Good luck!!

File: lab_1_classify_by_unigrams/main.py (hunk 17 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:30]: i don't understand what you mean

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:26]: this symbol is not a letter, so i decided to delete it

File: lab_1_classify_by_unigrams/main.py (hunk 17 above)
Comment (AnnaBeliavskaya) [reply lab_1_classify_by_unigrams/main.py:30]: done

**Pull request closed without merging**